src/uct/tcp/tcpcm/tcpcm_ep.c

```diff
@@ -0,0 +1,298 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017-2019.  ALL RIGHTS RESERVED.
+ * Copyright (C) NVIDIA Corporation. 2019.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#include "tcpcm_ep.h"
+
+
+#define UCT_TCPCM_CB_FLAGS_CHECK(_flags) \
+    do { \
+        UCT_CB_FLAGS_CHECK(_flags); \
+        if (!((_flags) & UCT_CB_FLAG_ASYNC)) { \
+            return UCS_ERR_UNSUPPORTED; \
+        } \
+    } while (0)
+
+ucs_status_t uct_tcpcm_ep_set_sock_id(uct_tcpcm_iface_t *iface, uct_tcpcm_ep_t *ep)
+{
+    ucs_status_t status;
+    struct sockaddr *dest_addr = NULL;
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+
+    /* create a sock_id for the client side */
+    if (iface->sock_id_quota > 0) {
+        /* Create an id for this interface. Events associated with this id will be
+         * reported on the event_channel that was created on iface init. */
+        ep->sock_id_ctx = ucs_malloc(sizeof(*ep->sock_id_ctx), "client sock_id_ctx");
+        if (ep->sock_id_ctx == NULL) {
+            status = UCS_ERR_NO_MEMORY;
+            goto out;
+        }
+
+        /* FIXME: Review */
+        dest_addr = (struct sockaddr *) &(ep->remote_addr);
+
+        status = ucs_socket_create(dest_addr->sa_family, SOCK_STREAM,
+                                   &ep->sock_id_ctx->sock_id);
+        if (status != UCS_OK) {
+            goto out_free;
+        }
+
+        ep->sock_id_ctx->ep = ep;
+        ucs_list_add_tail(&iface->used_sock_ids_list, &ep->sock_id_ctx->list);
+        iface->sock_id_quota--;
+        ucs_debug("ep %p, new sock_id %d. sock_id_in_quota %d", ep,
+                   ep->sock_id_ctx->sock_id, iface->sock_id_quota);
+        status = UCS_OK;
+        goto out;
+    } else {
+        ep->sock_id_ctx = NULL;
+        status = UCS_ERR_NO_RESOURCE;
+        goto out;
+    }
+
+out_free:
+    ucs_free(ep->sock_id_ctx);
+out:
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+    return status;
+}
+
+ucs_status_t uct_tcpcm_send_client_info(uct_tcpcm_iface_t *iface, uct_tcpcm_ep_t *ep)
+{
+    uct_tcpcm_conn_param_t conn_param;
+    uct_tcpcm_priv_data_hdr_t *hdr;
+    ssize_t sent_len = 0;
+    ssize_t recv_len = 0;
+    ssize_t offset = 0;
+    int connect_confirm = -1;
+
+    memset(&conn_param.private_data, 0, UCT_TCPCM_UDP_PRIV_DATA_LEN);
+    hdr = &conn_param.hdr;
+
+    /* pack worker address into private data */
+    hdr->length = ep->pack_cb(ep->pack_cb_arg, "eth0",
```

Comment:
Should we pass an address on which worker is bound instead of `"eth0"`?

Reply:
Can you point to the generic handle? So that I don't have to pass `eth0`.